### ISSUE TITLE
`invlink` and Reverse Mode AD - (1) method error for InverseWishart Distribution, (2) different solutions for ForwardDiff vs ReverseDiff/Zygote

### ISSUE THREAD

**Issue by paschermayr** [TuringLang/Bijectors.jl#119]
Hi guys, 

I am using the following package versions (updated to the current DistributionsAD module after last closed issue):

```
  [76274a88] Bijectors v0.8.0     
  [31c24e10] Distributions v0.23.4  
  [ced4e74d] DistributionsAD v0.6.1 
  [f6369f11] ForwardDiff v0.10.10 
  [37e2e3b7] ReverseDiff v1.2.0
  [e88e6eb3] Zygote v0.4.22 
```

I have a problem using `invlink` and Reverse Mode AD. When I only import `ForwardDiff `and `ReverseDiff`, there seems to be method error for the logpdf_with_trans evaluation for the `InverseWishart` distribution (all other distributions that I tested worked). MWE is here:

```
using Distributions, DistributionsAD, Bijectors
using ForwardDiff, ReverseDiff#, Zygote

function get_logpost(distr::Distributions.Distribution)
    function logposterior(θₜ::AbstractVector{<:Real})
        θ  = invlink(distr, reshape(θₜ, (2,2) ) )
        lp = logpdf_with_trans(distr, θ, true) #method error here when using ReverseDiff
        ll = logpdf(MvNormal([0., .0], θ), [2., 2.])
        return ll + lp
    end
end
theta_transformed = randn(4)
lposterior = get_logpost( InverseWishart(10, [0.15 0.05 ; 0.05 .15]) )
lposterior(theta_transformed)

ForwardDiff.gradient(lposterior, theta_transformed)
ReverseDiff.gradient(lposterior, theta_transformed) #error - MethodError: no method matching pullback(::DistributionsAD.ReverseDiff....
```

Now, if I additionally import Zygote, ReverseDiff gradient evaluation suddenly works. However, I get different solutions for ForwardDiff and ReverseDiff/Zygote unfortunately. I suspect the ForwardDiff solution to be correct, MWE here:

```
using Distributions, DistributionsAD, Bijectors
using ForwardDiff, ReverseDiff, Zygote

function get_logpost2(distr1::Distributions.Distribution, distr2::Distributions.Distribution)
    function logposterior2(θₜ::AbstractVector{<:Real})
        mu  = invlink(distr1, θₜ[1:2] )
        sigma  = invlink(distr2, reshape(θₜ[3:6], (2,2) ) )

        lp = logpdf_with_trans(distr1, mu, true)
        lp += logpdf_with_trans(distr2, sigma, true)

        ll = logpdf(MvNormal(mu, sigma), [2., 2.])

        return ll + lp
    end
end
theta_transformed = randn(6)
lposterior2 = get_logpost2( MvNormal([.1, .1]), InverseWishart(10, [0.15 0.05 ; 0.05 .15]) )
lposterior2(theta_transformed)

ForwardDiff.gradient(lposterior2, theta_transformed)
ReverseDiff.gradient(lposterior2, theta_transformed) #works now, but first 2 parameter with MvNormal prior have different solution.
Zygote.gradient(lposterior2, theta_transformed) #same solution as ReverseDiff
```

**Comment by devmotion** [comment]
The first issue (the method error) I can explain quite easily: DistributionsAD actually mixes different AD backends in some implementations (I assume because some computations are faster with Zygote and the other way around), and if you want to perform these computations you have to load both backends. That's what's happening here: in https://github.com/TuringLang/DistributionsAD.jl/blob/d7ceffbe3cd5ef54f98cdac6f39142ee9d3f8895/src/reversediffx.jl#L168-L170 you can see that the method for ReverseDiff actually uses a pullback from Zygote - but that one does not exist if you don't load Zygote.

Since the AD backends are all heavy dependencies, all AD functionality is hidden behind `@requires` blocks. The "correct" approach would be to put the implementation in a nested `@requires` block, such that the Zygote functionality is guaranteed to be available. However, that wouldn't actually fix your problem since then just the whole function (and not only the pullback) is not available anymore. So I'm not really sure what's the best approach to avoid these issues apart from not mixing the AD backends at all (which I assume would however lead to decreased performance in some cases).

BTW probably the mix of ReverseDiff and Zygote also explains why they both yield the same result.

**Comment by paschermayr** [comment]
Thank you very much for the information, that helps a lot! 

Regarding problem 2: When I set `get_logpost2( MvNormal([X=>1, Y=>1]), ...)`, I do get identical results for both ForwardDiff and ReverseDiff. This problem only arises as soon as `get_logpost2( MvNormal([X<1, Y<1]), ...)`.

**Comment by devmotion** [comment]
I noticed the following
```julia
julia> using DistributionsAD, Distributions, FiniteDifferences, ForwardDiff, Tracker, ReverseDiff, Zygote

julia> function h(theta, x)
           dist = MvNormal(theta)
           return logpdf(dist, x)
       end
h (generic function with 1 method)

julia> x = randn(2)
2-element Array{Float64,1}:
 -0.2728447604886671
  0.2954077831645062

julia> FiniteDifferences.grad(central_fdm(5, 1), x -> h([0.1, 0.1], x), x)[1]
2-element Array{Float64,1}:
  27.28447604886685
 -29.540778316450915

julia> ForwardDiff.gradient(x -> h([0.1, 0.1], x), x)
2-element Array{Float64,1}:
  27.284476048866708
 -29.540778316450616

julia> Tracker.gradient(x -> h([0.1, 0.1], x), theta)[1]
Tracked 2-element Array{Float64,1}:
  2728.4476048866704
 -2954.077831645061

julia> ReverseDiff.gradient(x -> h([0.1, 0.1], x), x)
2-element Array{Float64,1}:
  2728.4476048866704
 -2954.077831645061

julia> Zygote.gradient(x -> h([0.1, 0.1], x), x)[1]
2-element Array{Float64,1}:
  27.284476048866708
 -29.54077831645062
```
with
```julia
  [31c24e10] Distributions v0.23.4
  [26cc04aa] FiniteDifferences v0.10.2
  [f6369f11] ForwardDiff v0.10.10
  [37e2e3b7] ReverseDiff v1.2.0
  [9f7883ad] Tracker v0.2.7
  [e88e6eb3] Zygote v0.5.1
```
Moreover, I get
```julia
julia> FiniteDifferences.grad(central_fdm(5, 1), x -> h([10.0, 10.0], x), x)[1]
2-element Array{Float64,1}:
  0.0027284476047971617
 -0.0029540778316831193

julia> ForwardDiff.gradient(x -> h([10.0, 10.0], x), x)
2-element Array{Float64,1}:
  0.0027284476048866713
 -0.002954077831645062

julia> Tracker.gradient(x -> h([10.0, 10.0], x), theta)[1]
Tracked 2-element Array{Float64,1}:
  2.7284476048866714e-5
 -2.954077831645062e-5

julia> ReverseDiff.gradient(x -> h([10.0, 10.0], x), x)
2-element Array{Float64,1}:
  2.7284476048866714e-5
 -2.954077831645062e-5

julia> Zygote.gradient(x -> h([10.0, 10.0], x), x)[1]
2-element Array{Float64,1}:
  0.0027284476048866713
 -0.002954077831645062
```
so for me it fails for standard deviations < 1 and > 1 consistently for Tracker and ReverseDiff. Something seems to be broken there, I'll open an issue over at DistributionsAD.

**Comment by devmotion** [comment]
With https://github.com/TuringLang/DistributionsAD.jl/pull/92, FiniteDifferences, Tracker, ForwardDiff, and ReverseDiff all yield the same gradient. Seems Zygote is still broken though, not sure why yet.

**Comment by devmotion** [comment]
The latest commit in https://github.com/TuringLang/DistributionsAD.jl/pull/92 fixes Zygote as well, so the second part of this issue here is resolved.

**Comment by devmotion** [comment]
The AD issues are fixed in DistributionsAD 0.6.2.
